admin/server/projects.go

```diff
@@ -247,94 +247,112 @@ func (s *Server) GetProject(ctx context.Context, req *adminv1.GetProjectRequest)
-
 		var c *client.Client
+		if mdl.TransitiveAccessResource != nil {
+			c, err = s.admin.OpenRuntimeClient(depl)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "could not open runtime client: %s", err.Error())
+			}
+			defer c.Close() // nolint:gocritic // client is created only once
 
-		for _, r := range mdl.Resources {
-			condition.WriteString(fmt.Sprintf(" OR ('{{.self.kind}}'=%s AND '{{lower .self.name}}'=%s)", duckdbsql.EscapeStringValue(r.Type), duckdbsql.EscapeStringValue(strings.ToLower(r.Name))))
+			rules = append(rules, &runtimev1.SecurityRule{
+				Rule: &runtimev1.SecurityRule_TransitiveAccess{
+					TransitiveAccess: &runtimev1.SecurityRuleTransitiveAccess{
+						Resource: &runtimev1.ResourceName{
+							Kind: mdl.TransitiveAccessResource.Type,
+							Name: mdl.TransitiveAccessResource.Name,
+						},
+					},
+				},
+			})
+		} else {
+			// Build condition for what the magic auth token can access
+			var condition strings.Builder
+			// All themes
+			condition.WriteString(fmt.Sprintf("'{{.self.kind}}'='%s'", runtime.ResourceKindTheme))
+
+			for _, r := range mdl.Resources {
+				condition.WriteString(fmt.Sprintf(" OR ('{{.self.kind}}'=%s AND '{{lower .self.name}}'=%s)", duckdbsql.EscapeStringValue(r.Type), duckdbsql.EscapeStringValue(strings.ToLower(r.Name))))
+
+				// If the magic token's resource is an Explore, we also need to include its underlying metrics view
+				if r.Type == runtime.ResourceKindExplore {
+					if c == nil {
+						c, err = s.admin.OpenRuntimeClient(depl)
+						if err != nil {
+							return nil, status.Errorf(codes.Internal, "could not open runtime client: %s", err.Error())
+						}
+						defer c.Close() // nolint:gocritic // client is created only once
+					}
 
-			// If the magic token's resource is an Explore, we also need to include its underlying metrics view
-			if r.Type == runtime.ResourceKindExplore {
-				if c == nil {
-					c, err = s.admin.OpenRuntimeClient(depl)
+					resp, err := c.GetResource(ctx, &runtimev1.GetResourceRequest{
+						InstanceId: depl.RuntimeInstanceID,
+						Name: &runtimev1.ResourceName{
+							Kind: r.Type,
+							Name: r.Name,
+						},
+					})
 					if err != nil {
-						return nil, status.Errorf(codes.Internal, "could not open runtime client: %s", err.Error())
+						if status.Code(err) == codes.NotFound {
+							return nil, status.Errorf(codes.NotFound, "resource for magic token not found (name=%q, type=%q)", r.Name, r.Type)
+						}
+						return nil, fmt.Errorf("could not get resource for magic token: %w", err)
 					}
-					defer c.Close() // nolint:gocritic // client is created only once
+
+					spec := resp.Resource.GetExplore().State.ValidSpec
+					if spec != nil {
+						condition.WriteString(fmt.Sprintf(" OR ('{{.self.kind}}'='%s' AND '{{lower .self.name}}'=%s)", runtime.ResourceKindMetricsView, duckdbsql.EscapeStringValue(strings.ToLower(spec.MetricsView))))
+					}
```

Comment:
This handling would not be needed anymore if we always give transitive access to all the `Resources`, right? I.e. old public URLs and report tokens would just get the metrics access through the transitive rule instead, right?

Reply:
> we will still need for old magic tokens to keep backwards compatibility?

You will have to check, I don't remember, but I think the old magic tokens have the explore name in the `resources` already, so transitive access would just work?